File: src/translate/openai.ts
import crypto from "node:crypto";
import { toolArgumentSuffix } from "./tool-arguments.js";
import type {
  OpenAIChatRequest,
  OpenAIMessage,
  OpenAITool,
  ContentPart,
  ToolChoice,
  ToolCall,
  CCMessage,
  CCContentPart,
  CCTool,
  CCRequestBody,
  CCToolChoice,
  CCEvent,
  UsageData,
} from "@/translate/types.js";
import { resolveModel, resolveEffortForModel } from "@/translate/models.js";
import { tagStreamError, UpstreamEventError } from "@/stream.js";
import {
  applyNoToolsSafeguard,
  extractUsage,
  pruneDanglingTools,
  buildCCConfig,
} from "@/translate/util.js";
import { logger } from "@/logger.js";

// ──────────────────────────────────────────
// OpenAI request → CC request
// ──────────────────────────────────────────

function toCCMessages(messages: OpenAIMessage[]): {
  ccMessages: CCMessage[];
  systemPrompt: string | undefined;
} {
  const systemParts: string[] = [];
  const ccMessages: CCMessage[] = [];

  // Map tool_call_id -> toolName so tool-result parts can carry the tool name
  // required by the CC upstream schema (tool-result parts need a `toolName`).
  const toolNameById = new Map<string, string>();
  for (const msg of messages) {
    if (msg.tool_calls) {
      for (const tc of msg.tool_calls) {
        if (tc.id) toolNameById.set(tc.id, tc.function.name);
      }
    }
  }

  for (const msg of messages) {
    if (msg.role === "system" || msg.role === "developer") {
      const text =
        typeof msg.content === "string"
          ? msg.content
          : msg.content
              .map((p) => (p.type === "text" ? p.text : ""))
              .filter(Boolean)
              .join("\n");
      systemParts.push(text);
      continue;
    }

    if (msg.role === "tool") {
      const resultText =
        typeof msg.content === "string" ? msg.content : JSON.stringify(msg.content);
      ccMessages.push({
        role: "tool",
        content: [
          {
            type: "tool-result",
            toolCallId: msg.tool_call_id ?? "",
            toolName: (msg.tool_call_id ? toolNameById.get(msg.tool_call_id) : undefined) ?? "",
            output: { type: "text", value: resultText },
          },
        ],
      });
      continue;
    }

    if (msg.role === "assistant" && msg.tool_calls) {
      const parts: CCContentPart[] = [];
      if (typeof msg.content === "string" && msg.content) {
        parts.push({ type: "text", text: msg.content });
      }
      for (const tc of msg.tool_calls) {
        let parsedInput: unknown;
        try {
          parsedInput = tc.function.arguments ? JSON.parse(tc.function.arguments) : {};
        } catch {
          parsedInput = tc.function.arguments;
        }
        parts.push({
          type: "tool-call",
          toolCallId: tc.id,
          toolName: tc.function.name,
          input: parsedInput,
        });
      }
      ccMessages.push({ role: "assistant", content: parts });
      continue;
    }

    if (msg.role === "user") {
      if (typeof msg.content === "string") {
        ccMessages.push({ role: "user", content: msg.content });
      } else {
        const parts: CCContentPart[] = msg.content.map((p: ContentPart) => {
          if (p.type === "image_url" && p.image_url) {
            return { type: "image", image: p.image_url.url };
          }
          return { type: "text", text: p.text ?? "" };
        });
        ccMessages.push({ role: "user", content: parts });
      }
      continue;
    }

    // assistant (plain text)
    if (typeof msg.content === "string") {
      ccMessages.push({ role: "assistant", content: msg.content });
    }
  }

  return {
    ccMessages: pruneDanglingTools(ccMessages),
    systemPrompt: systemParts.length > 0 ? systemParts.join("\n\n") : undefined,
  };
}

function toCCTools(tools: OpenAITool[] | undefined): CCTool[] | undefined {
  if (!tools || tools.length === 0) return undefined;
  return tools.map((t) => ({
    name: t.function.name,
    description: t.function.description,
    input_schema: t.function.parameters,
  }));
}

/**
 * Map an OpenAI `tool_choice` to the object form CC's bridge requires.
 * CC validates `params.tool_choice` as an object and rejects bare strings.
 * Accepted types: "auto" | "any" | "tool". There is no "required" (use "any")
 * and no "none" — "none" has no CC equivalent, so we fall back to the default
 * (omit) and let the model decide.
 */
function resolveToolChoice(tc: ToolChoice | undefined): CCToolChoice | undefined {
  if (!tc || tc === "auto" || tc === "none") return undefined;
  if (tc === "required") return { type: "any" };
  if (typeof tc === "object" && tc.type === "function") {
    return { type: "tool", name: tc.function.name };
  }
  return undefined;
}

export function toCCRequest(
  req: OpenAIChatRequest,
  configOverrides?: Partial<CCRequestBody["config"]>,
): CCRequestBody {
  const { ccMessages, systemPrompt } = toCCMessages(req.messages);
  const resolvedModel = resolveModel(req.model);

  const body: CCRequestBody = {
    config: buildCCConfig(configOverrides),
    memory: "",
    taste: "",
    skills: "",
    permissionMode: "standard",
    params: {
      model: resolvedModel,
      messages: ccMessages,
      stream: req.stream ?? false,
      max_tokens: req.max_tokens,
      temperature: req.temperature,
      top_p: req.top_p,
      stop: req.stop,
      reasoning_effort: resolveEffortForModel(resolvedModel, req.reasoning_effort),
      tools: toCCTools(req.tools),
      tool_choice: resolveToolChoice(req.tool_choice),
    },
    threadId: crypto.randomUUID(),
  };

  if (systemPrompt) body.params.system = systemPrompt;
  applyNoToolsSafeguard(body, ccMessages, req.tools != null && req.tools.length > 0);

  return body;
}

// ──────────────────────────────────────────
// CC events → OpenAI streaming chunks
// ──────────────────────────────────────────

const OPENAI_FINISH_MAP: Record<string, string> = {
  stop: "stop",
  length: "length",
  content_filtered: "content_filter",
  "tool-call": "tool_calls",
  "tool-calls": "tool_calls",
  tool_call: "tool_calls",
  error: "stop",
};

function toOpenAIUsage(u: UsageData): Record<string, unknown> {
  const promptTokens = u.promptTokens ?? 0;
  const completionTokens = u.completionTokens ?? 0;
  const usage: Record<string, unknown> = {
    prompt_tokens: promptTokens,
    completion_tokens: completionTokens,
    total_tokens: u.totalTokens ?? promptTokens + completionTokens,
  };
  if (u.promptTokensDetails?.cachedTokens != null) {
    usage.prompt_tokens_details = { cached_tokens: u.promptTokensDetails.cachedTokens };
  }
  if (u.completionTokensDetails?.reasoningTokens != null) {
    usage.completion_tokens_details = {
      reasoning_tokens: u.completionTokensDetails.reasoningTokens,
    };
  }
  return usage;
}

export class OpenAIStreamEncoder {
  readonly id: string;
  private readonly created: number;
  private toolCallIndex = 0;
  private sawFinish = false;
  /**
   * Whether any real payload (text, reasoning, or a tool call) has been sent.
   *
   * The opening role chunk does not count: it carries no payload and clients
   * merge a repeated one silently, so a retry after only that chunk still
   * presents a single clean response. This is what decides whether an upstream
   * `error` event is recoverable by re-sending the request.
   */
  private emittedContent = false;
  /** finish 事件里归一后的用量（含缓存命中），供服务层落统计日志 */
  lastUsage?: UsageData;
  // Map a CC toolCallId to the OpenAI streaming `index` we assigned it.
  // Without this, parallel tool-call-delta streams without an explicit
  // `index` field would all be assigned index 0 and the client would merge
  // them into a single tool call.
  private readonly toolCallIdToIndex = new Map<string, number>();
  private readonly toolArguments = new Map<number, string>();
  private readonly toolMetadata = new Map<number, { id?: string; name?: string }>();

  private toolMetadataDelta(index: number, id?: string, name?: string) {
    const seen = this.toolMetadata.get(index) ?? {};
    const delta: { id?: string; type?: string; name?: string } = {};
    for (const [field, value] of [["id", id], ["name", name]] as const) {
      if (!value) continue;
      if (seen[field] && seen[field] !== value) {
        throw new Error("Inconsistent upstream tool metadata");
      }
      if (!seen[field]) {
        seen[field] = value;
        delta[field] = value;
        if (field === "id") delta.type = "function";
      }
    }
    this.toolMetadata.set(index, seen);
    return delta;
  }

  constructor(private readonly model: string) {
    this.id = crypto.randomUUID();
    this.created = Math.floor(Date.now() / 1000);
  }

  get finished(): boolean {
    return this.sawFinish;
  }

  /**
   * Whether the client has received anything worth keeping.
   *
   * False means a failed attempt can be re-sent invisibly — at most the
   * opening role chunk went out, which clients merge without complaint.
   */
  get hasEmittedContent(): boolean {
    return this.emittedContent;
  }

  /**
   * Resolve the OpenAI streaming `index` for a given tool-call id, allocating
   * a new one on first sighting. Falls back to the upstream-provided index
   * (if any) so we don't fight a bridge that already numbers them.
   */
  private resolveToolCallIndex(
    toolCallId: string | undefined,
    upstreamIndex: number | undefined,
  ): number {
    if (toolCallId) {
      const known = this.toolCallIdToIndex.get(toolCallId);
      if (known !== undefined) return known;
      const idx = upstreamIndex ?? this.toolCallIndex++;
      this.toolCallIdToIndex.set(toolCallId, idx);
      return idx;
    }
    return upstreamIndex ?? this.toolCallIndex++;
  }

  emit(event: CCEvent): object[] {
    // 终态守卫：finish/error 已收尾（finish chunk 或 error 信封），其后上游
    // 再发的一切事件到此吞掉，防止重复 finish 或终态后的内容块。
    if (this.sawFinish) return [];
    const chunks: object[] = [];
    const id = this.id;
    const created = this.created;

    switch (event.type) {
      case "start": {
        this.toolCallIndex = 0;
        chunks.push({
          id,
          object: "chat.completion.chunk",
          created,
          model: this.model,
          choices: [{ index: 0, delta: { role: "assistant" }, finish_reason: null }],
        });
        break;
      }

      case "text-delta": {
        const text = event.data.text as string;
        if (text) {
          this.emittedContent = true;
          chunks.push({
            id,
            object: "chat.completion.chunk",
            created,
            model: this.model,
            choices: [{ index: 0, delta: { content: text }, finish_reason: null }],
          });
        }
        break;
      }

      case "reasoning-delta": {
        const text = event.data.text as string;
        if (text) {
          this.emittedContent = true;
          chunks.push({
            id,
            object: "chat.completion.chunk",
            created,
            model: this.model,
            choices: [{ index: 0, delta: { reasoning_content: text }, finish_reason: null }],
          });
        }
        break;
      }

      case "tool-call-delta": {
        this.emittedContent = true;
        const toolCallId = (event.data.toolCallId as string) ?? undefined;
        const upstreamIndex = (event.data.index as number) ?? undefined;
        const tc: {
          index: number;
          id?: string;
          type?: string;
          function: { name?: string; arguments: string };
        } = {
          index: this.resolveToolCallIndex(toolCallId, upstreamIndex),
          function: { arguments: (event.data.arguments as string) ?? "" },
        };
        const metadata = this.toolMetadataDelta(tc.index, toolCallId, event.data.name as string | undefined);
        if (metadata.id) {
          tc.id = metadata.id;
          tc.type = metadata.type;
        }
        if (metadata.name) {
          tc.function.name = metadata.name;
        }
        this.toolArguments.set(
          tc.index,
          (this.toolArguments.get(tc.index) ?? "") + tc.function.arguments,
        );
        chunks.push({
          id,
          object: "chat.completion.chunk",
          created,
          model: this.model,
          choices: [{ index: 0, delta: { tool_calls: [tc] }, finish_reason: null }],
        });
        break;
      }

      case "tool-call": {
        this.emittedContent = true;
        const toolCallId = (event.data.toolCallId as string) ?? "";
        const toolName = (event.data.toolName as string) ?? (event.data.name as string) ?? "";
        const input = event.data.input ?? event.data.arguments;
        const args = typeof input === "string" ? input : input != null ? JSON.stringify(input) : "";
        // Reuse the index allocated by any prior tool-call-delta for this id
        // so the client merges the final tool-call with the streamed deltas
        // instead of seeing it as a brand-new tool call.
        const index = this.resolveToolCallIndex(
          toolCallId || undefined,
          (event.data.index as number) ?? undefined,
        );
        const emitted = this.toolArguments.get(index) ?? "";
        const suffix = toolArgumentSuffix(emitted, args);
        this.toolArguments.set(index, emitted + suffix);
        const { name, ...identity } = this.toolMetadataDelta(index, toolCallId, toolName);
        chunks.push({
          id,
          object: "chat.completion.chunk",
          created,
          model: this.model,
          choices: [
            {
              index: 0,
              delta: {
                tool_calls: [
                  {
                    index,
                    ...identity,
                    function: { ...(name ? { name } : {}), arguments: suffix },
                  },
                ],
              },
              finish_reason: null,
            },
          ],
        });
        break;
      }

      case "finish": {
        this.sawFinish = true;
        const usage = extractUsage(event.data as Record<string, unknown>);
        this.lastUsage = usage;
        const finishReason = (event.data.finishReason as string) ?? "stop";
        chunks.push({
          id,
          object: "chat.completion.chunk",
          created,
          model: this.model,
          choices: [
            { index: 0, delta: {}, finish_reason: OPENAI_FINISH_MAP[finishReason] ?? "stop" },
          ],
        });
        if (usage) {
          chunks.push({
            id,
            object: "chat.completion.chunk",
            created,
            model: this.model,
            choices: [],
            usage: toOpenAIUsage(usage),
          });
        }
        break;
      }

      case "error": {
        const message =
          (event.data.message as string) ??
          (event.data.error as { message?: string } | undefined)?.message ??
          JSON.stringify(event.data);
        // Nothing meaningful written yet → recoverable: throw so the service
        // layer can re-send upstream and the client sees one clean response.
        // The role chunk alone does not count as output (clients merge a
        // repeated role delta without complaint — measured).
        if (!this.emittedContent) throw new UpstreamEventError(message);
        this.sawFinish = true;
        return this.errorChunks(message);
      }
    }

    return chunks;
  }

  /**
   * Build the OpenAI error envelope for a failed generation. Used both by
   * `emit()` for upstream error events and by the server's pumpStream catch
   * for stream-level errors (TCP failure, idle timeout, encoder throw).
   *
   * The envelope — not a content chunk — is what signals failure. The AI SDK's
   * OpenAI-compatible chunk schema is a union of the normal chunk shape and
   * `{error: {...}}`; `doStream` turns the error arm into a stream error part
   * and sets finishReason "error", which is what lets the caller retry. Wrapping
   * the message in delta.content instead makes the client read the failure as a
   * normal assistant reply and report a successful turn, so the error is
   * silently swallowed.
   */
  private errorChunks(message: string): object[] {
    logger.error(`[CC upstream error] ${message}`);
    return [
      {
        error: {
          message,
          type: "upstream_error",
          // code:"network_error" 是下游 ZCode 重试分类器的识别字：命中即判
          // retryable:true（NetworkError），否则 business error 一律不重试，
          // 上游 11 次重试额度全弃。type 保留 upstream_error 供人类排查。
          code: "network_error",
        },
      },
    ];
  }

  errorChunk(err: Error): object {
    // Retained for backwards-compat with tests; new code should call
    // errorChunks() instead so the error appears inside a valid chunk.
    return {
      error: {
        message: err.message,
        type: "upstream_error",
        code: "network_error",
      },
    };
  }

  /**
   * Build the closing chunks for a stream that ended without a `finish`
   * event from the upstream (e.g. upstream connection dropped mid-response).
   * Emits a synthetic finish_reason so the client SDK doesn't see a
   * truncated response.
   */
  finishChunks(reason: string = "stop"): object[] {
    return [
      {
        id: this.id,
        object: "chat.completion.chunk",
        created: this.created,
        model: this.model,
        choices: [{ index: 0, delta: {}, finish_reason: reason }],
      },
    ];
  }

  /**
   * Public wrapper around errorChunks() for stream-level errors caught by
   * pumpStream (TCP failure, idle timeout, encoder throw). Tags the message
   * with the failure's real origin so the cause stays visible.
   */
  streamErrorChunks(err: Error): object[] {
    return this.errorChunks(tagStreamError(err));
  }
}

// ──────────────────────────────────────────
// CC events → OpenAI non-streaming response
// ──────────────────────────────────────────

interface FinishEvent {
  finishReason?: string;
  usage?: UsageData;
}

export function buildNonStreamingResponse(events: CCEvent[], model: string, id: string): object {
  let content = "";
  let reasoningContent = "";
  const toolCalls: ToolCall[] = [];
  let finish: FinishEvent | null = null;

  for (const event of events) {
    switch (event.type) {
      case "error":
        // Do not turn failed generations (or private diagnostics) into content.
        throw new Error("CC upstream generation failed");
      case "text-delta":
        content += (event.data.text as string) ?? "";
        break;
      case "reasoning-delta":
        reasoningContent += (event.data.text as string) ?? "";
        break;
      case "tool-call-delta": {
        const existing = toolCalls.find((tc) => tc.id === (event.data.toolCallId as string));
        if (existing) {
          existing.function.arguments += (event.data.arguments as string) ?? "";
        } else {
          toolCalls.push({
            id: (event.data.toolCallId as string) ?? "",
            type: "function",
            function: {
              name: (event.data.name as string) ?? "",
              arguments: (event.data.arguments as string) ?? "",
            },
          });
        }
        break;
      }
      case "tool-call": {
        // CC bridges commonly emit tool-call-delta * N then a final tool-call
        // with the same id. If we already accumulated args from deltas,
        // replace that entry with the canonical final payload (matches what
        // the streaming encoder does). Otherwise this is a one-shot tool
        // call with no preceding deltas — push it.
        const id = (event.data.toolCallId as string) ?? "";
        const input = event.data.input ?? event.data.arguments;
        const argsStr =
          typeof input === "string" ? input : input != null ? JSON.stringify(input) : "";
        const name = (event.data.toolName as string) ?? (event.data.name as string) ?? "";
        const existingIdx = toolCalls.findIndex((tc) => tc.id === id);
        const entry: ToolCall = {
          id,
          type: "function",
          function: { name, arguments: argsStr },
        };
        if (existingIdx >= 0) {
          toolCalls[existingIdx] = entry;
        } else {
          toolCalls.push(entry);
        }
        break;
      }
      case "finish":
        finish = {
          finishReason: event.data.finishReason as string,
          usage: extractUsage(event.data),
        };
        break;
    }
  }

  const message: Record<string, unknown> = { role: "assistant" };
  if (reasoningContent) message.reasoning_content = reasoningContent;
  if (toolCalls.length > 0) message.tool_calls = toolCalls.slice();
  if (content) {
    message.content = content;
  } else if (toolCalls.length === 0) {
    message.content = "";
  }

  const finishReason = finish?.finishReason ?? "stop";

  const response: Record<string, unknown> = {
    id,
    object: "chat.completion",
    created: Math.floor(Date.now() / 1000),
    model,
    choices: [
      {
        index: 0,
        message,
        finish_reason: OPENAI_FINISH_MAP[finishReason] ?? "stop",
      },
    ],
  };

  if (finish?.usage) response.usage = toOpenAIUsage(finish.usage);

  return response;
}

File: crates/ccproxy/src/translate/openai.rs
//! OpenAI chat completions -> CC `/alpha/generate` body, ported from
//! src/translate/openai.ts. Only the request half lives here; the streaming
//! encoder is M3.

use crate::models::Catalog;
use crate::translate::models::{resolve_effort_for_model, resolve_model, ModelTables};
use crate::translate::util::{
    apply_no_tools_safeguard, build_cc_config, prune_dangling_tools, truthy,
};
use serde_json::{json, Map, Value};
use std::collections::HashMap;

/// Build the CC body for an OpenAI chat request. Optional fields that the
/// client did not send are omitted rather than nulled, exactly as
/// `JSON.stringify` drops `undefined`.
///
/// `guard_off` carries the `CC_NO_TOOLS_GUARD=off` opt-out so the environment
/// is read by the caller (see `Env::guard_off`), keeping this a pure function.
pub fn to_cc_request(
    req: &Value,
    catalog: &Catalog,
    tables: &ModelTables,
    guard_off: bool,
) -> Value {
    let messages: Vec<Value> = req
        .get("messages")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();

    let (mut cc_messages, system_prompt) = to_cc_messages(&messages);
    prune_dangling_tools(&mut cc_messages);

    let requested_model = req.get("model").and_then(Value::as_str).unwrap_or("");
    let resolved_model = resolve_model(requested_model, catalog, tables);

    let mut params = Map::new();
    params.insert("model".into(), Value::String(resolved_model.clone()));
    params.insert("messages".into(), Value::Array(cc_messages));
    params.insert(
        "stream".into(),
        match req.get("stream") {
            None | Some(Value::Null) => Value::Bool(false),
            Some(v) => v.clone(),
        },
    );
    for key in ["max_tokens", "temperature", "top_p", "stop"] {
        if let Some(v) = req.get(key) {
            params.insert(key.into(), v.clone());
        }
    }
    if let Some(effort) = resolve_effort_for_model(
        &resolved_model,
        req.get("reasoning_effort").and_then(Value::as_str),
        tables,
    ) {
        params.insert("reasoning_effort".into(), Value::String(effort));
    }
    if let Some(tools) = to_cc_tools(req.get("tools")) {
        params.insert("tools".into(), tools);
    }
    if let Some(tc) = resolve_tool_choice(req.get("tool_choice")) {
        params.insert("tool_choice".into(), tc);
    }
    // Set before the safeguard so the instruction appends to the client's own.
    if let Some(sp) = system_prompt {
        if !sp.is_empty() {
            params.insert("system".into(), Value::String(sp));
        }
    }

    let has_tools = req
        .get("tools")
        .and_then(Value::as_array)
        .is_some_and(|t| !t.is_empty());
    apply_no_tools_safeguard(&mut params, has_tools, guard_off);

    json!({
        "config": build_cc_config(),
        "memory": "",
        "taste": "",
        "skills": "",
        "permissionMode": "standard",
        "params": Value::Object(params),
        "threadId": uuid::Uuid::new_v4().to_string(),
    })
}

/// Returns the CC messages plus the joined system prompt (`None` when the
/// client sent no system/developer message at all — an empty join counts as
/// absent, which is why the caller re-checks emptiness).
fn to_cc_messages(messages: &[Value]) -> (Vec<Value>, Option<String>) {
    // tool_call_id -> toolName: CC's tool-result parts must carry a toolName.
    let mut tool_name_by_id: HashMap<String, String> = HashMap::new();
    for msg in messages {
        if let Some(calls) = msg.get("tool_calls").and_then(Value::as_array) {
            for tc in calls {
                if let Some(id) = tc.get("id").and_then(Value::as_str) {
                    if let Some(name) = tc.pointer("/function/name").and_then(Value::as_str) {
                        tool_name_by_id.insert(id.to_string(), name.to_string());
                    }
                }
            }
        }
    }

    let mut system_parts: Vec<String> = Vec::new();
    let mut cc: Vec<Value> = Vec::new();

    for msg in messages {
        let role = msg.get("role").and_then(Value::as_str).unwrap_or("");
        let content = msg.get("content");

        match role {
            "system" | "developer" => {
                let text = match content {
                    Some(Value::String(s)) => s.clone(),
                    Some(Value::Array(parts)) => parts
                        .iter()
                        .map(|p| {
                            if p.get("type").and_then(Value::as_str) == Some("text") {
                                p.get("text")
                                    .and_then(Value::as_str)
                                    .unwrap_or("")
                                    .to_string()
                            } else {
                                String::new()
                            }
                        })
                        .filter(|s| !s.is_empty())
                        .collect::<Vec<_>>()
                        .join("\n"),
                    _ => String::new(),
                };
                system_parts.push(text);
            }

            "tool" => {
                let tool_call_id = match msg.get("tool_call_id") {
                    None | Some(Value::Null) => Value::String(String::new()),
                    Some(v) => v.clone(),
                };
                let tool_name = match tool_call_id.as_str() {
                    Some(id) if !id.is_empty() => tool_name_by_id
                        .get(id)
                        .map(String::as_str)
                        .unwrap_or("")
                        .to_string(),
                    _ => String::new(),
                };

                let mut output = Map::new();
                output.insert("type".into(), Value::String("text".into()));
                match content {
                    Some(Value::String(s)) => {
                        output.insert("value".into(), Value::String(s.clone()));
                    }
                    // Non-string content is JSON-stringified, not flattened.
                    Some(v) => {
                        output.insert(
                            "value".into(),
                            Value::String(serde_json::to_string(v).unwrap_or_default()),
                        );
                    }
                    // `JSON.stringify(undefined)` drops the key entirely.
                    None => {}
                }

                cc.push(json!({
                    "role": "tool",
                    "content": [{
                        "type": "tool-result",
                        "toolCallId": tool_call_id,
                        "toolName": tool_name,
                        "output": Value::Object(output),
                    }],
                }));
            }

            // Note the array-content case: an assistant message with tool_calls
            // becomes parts, while one without falls through to the plain-text
            // branch below and is dropped when its content is not a string.
            "assistant" if msg.get("tool_calls").is_some_and(Value::is_array) => {
                let mut parts: Vec<Value> = Vec::new();
                if let Some(text) = content.and_then(Value::as_str) {
                    if !text.is_empty() {
                        parts.push(json!({"type": "text", "text": text}));
                    }
                }
                if let Some(calls) = msg.get("tool_calls").and_then(Value::as_array) {
                    for tc in calls {
                        let mut part = Map::new();
                        part.insert("type".into(), Value::String("tool-call".into()));
                        if let Some(id) = tc.get("id") {
                            part.insert("toolCallId".into(), id.clone());
                        }
                        if let Some(name) = tc.pointer("/function/name") {
                            part.insert("toolName".into(), name.clone());
                        }
                        let args = tc
                            .pointer("/function/arguments")
                            .and_then(Value::as_str)
                            .unwrap_or("");
                        let input = if args.is_empty() {
                            json!({})
                        } else {
                            serde_json::from_str::<Value>(args)
                                .unwrap_or_else(|_| Value::String(args.to_string()))
                        };
                        part.insert("input".into(), input);
                        parts.push(Value::Object(part));
                    }
                }
                cc.push(json!({"role": "assistant", "content": parts}));
            }

            "user" => match content {
                Some(Value::String(s)) => cc.push(json!({"role": "user", "content": s})),
                Some(Value::Array(parts)) => {
                    let mapped: Vec<Value> = parts
                        .iter()
                        .map(|p| {
                            let is_image = p.get("type").and_then(Value::as_str)
                                == Some("image_url")
                                && p.get("image_url").is_some_and(truthy);
                            if is_image {
                                let mut m = Map::new();
                                m.insert("type".into(), Value::String("image".into()));
                                if let Some(url) = p.pointer("/image_url/url").filter(|u| truthy(u))
                                {
                                    m.insert("image".into(), url.clone());
                                }
                                Value::Object(m)
                            } else {
                                json!({
                                    "type": "text",
                                    "text": p.get("text").and_then(Value::as_str).unwrap_or(""),
                                })
                            }
                        })
                        .collect();
                    cc.push(json!({"role": "user", "content": mapped}));
                }
                _ => {}
            },

            "assistant" => {
                if let Some(Value::String(s)) = content {
                    cc.push(json!({"role": "assistant", "content": s}));
                }
            }

            _ => {}
        }
    }

    let system_prompt = if system_parts.is_empty() {
        None
    } else {
        Some(system_parts.join("\n\n"))
    };
    (cc, system_prompt)
}

/// OpenAI tools -> CC tools. An empty list is omitted (unlike the Anthropic
/// path, which forwards it).
fn to_cc_tools(tools: Option<&Value>) -> Option<Value> {
    let arr = tools?.as_array()?;
    if arr.is_empty() {
        return None;
    }
    let out: Vec<Value> = arr
        .iter()
        .map(|t| {
            let mut m = Map::new();
            for (from, to) in [
                ("/function/name", "name"),
                ("/function/description", "description"),
                ("/function/parameters", "input_schema"),
            ] {
                if let Some(v) = t.pointer(from) {
                    m.insert(to.into(), v.clone());
                }
            }
            Value::Object(m)
        })
        .collect();
    Some(Value::Array(out))
}

/// CC validates `tool_choice` as an object and rejects bare strings. There is
/// no "required" (use "any") and no "none" — "none" has no CC equivalent, so it
/// is omitted and the model decides.
fn resolve_tool_choice(tc: Option<&Value>) -> Option<Value> {
    let tc = tc?;
    if let Some(s) = tc.as_str() {
        return match s {
            "required" => Some(json!({"type": "any"})),
            _ => None,
        };
    }
    if tc.get("type").and_then(Value::as_str) == Some("function") {
        let mut m = Map::new();
        m.insert("type".into(), Value::String("tool".into()));
        if let Some(name) = tc.pointer("/function/name") {
            m.insert("name".into(), name.clone());
        }
        return Some(Value::Object(m));
    }
    None
}
